### DISCUSSION TITLE
Database Schema

### DISCUSSION THREAD

**Discussion by TrystonMinsquero** [UTDNebula/nebula-api#57] [Ideas] [answered]
Thread for all database schema finalization.

**Comment by jpahm** [comment]
No documentation to go with it yet, so no cookie for me, but I think I've put together [a good foundation](https://pastebin.com/UETCCkw2) that we can throw the documentation on top of. I know @jmyrick02 has said he has a good deal of documentation done already, so perhaps he can apply/adjust that to fit this?

I'm open to changing anything I've provided in the above document, but I think it's adherent to the agreements we've come to.

**Reply by jpahm** [reply]
*Also note I've trimmed the `_id` property from all of the types that do not need to be stored independently in the database, such as `Location`s or `Requirement`s

**Reply by TrystonMinsquero** [reply]
Could you post that on this thread and organize it similar to #10 please.

**Reply by jmyrick02** [reply]
I'm currently working on the documentation in the format of #10.

**Comment by jmyrick02** [comment, marked as answer]
# Course
## Overview

The `Course` object represents a course available at the University of Texas at Dallas. A `Course` should not be confused with a `Section` which is the actual instantiation of a `Course` with a professor and dedicated meeting times.

## Object Representation
```ts
Course = {
    "_id": ObjectId,
    "course_number": string,
    "subject_prefix": string,
    "title": string,
    "description": string,
    "school": string,
    "credit_hours": string,
    "class_level": string,
    "activity_type": string,
    "grading": string,
    "internal_course_number": string,
    "prerequisites": CollectionRequirement,
    "corequisites": CollectionRequirement,
    "lecture_contact_hours": string,
    "laboratory_contact_hours": string,
    "offering_frequency": string,
    "attributes": Object,
}
```

## Properties
> `._id`
> 
> **Type**: ObjectId
>
> The MongoDB database id for the `Course` object.
>
> **Example**: ObjectId("61ebbb126e3659537e8a14d6")

> `.course_number`
>
> **Type**: string
>
> The course's official number.
>
> **Example**: 2417

> `.subject_prefix`
>
> **Type**: string
>
> The course's subject prefix.
>
> **Example**: MATH

> `.title`
>
> **Type**: string
>
> The course's title.
>
> **Example**: Calculus I

> `.description`
>
> **Type**: string
>
> The course's description.
>
> **Example**: "Functions, limits, continuity, differentiation; integration of..."

> `.school`
>
> **Type**: string
>
> 
>
> **Example**: 

> `.credit_hours`
>
> **Type**: string
>
> The number of credit hours awarded by successful completion of the course. Will be "V" if variable credit hours.
>
> **Example**: "4", "V"

> `.class_level`
>
> **Type**: string
>
> The level of education that this course course corresponds to
>
> **Example**: 4

> `.activity_type`
>
> **Type**: string
>
> The type of class this course corresponds to.
>
> **Example**: Lecture

> `.grading`
>
> **Type**: string
>
> The grading status of this course.
>
> **Example**: Graded

> `.internal_course_number`
>
> **Type**: string
>
> The internal (university) number used to reference this course.
>
> **Example**: 008613

> `.prerequisites`
>
> **Type**: CollectionRequirement
>
> A collection of prerequisites that must be met before a student may enroll in a section of this course.

> `.corequisites`
>
> **Type**: CollectionRequirement
>
> A collection of all course requirements that must be met either before or while a student enrolls in a section of this course.

> `.lecture_contact_hours`
>
> **Type**: string
>
> The weekly contact hours in lecture for a course. This information is outlined in the UTD Course Policies page.
>
> **Example**: 2

> `.laboratory_contact_hours`
>
> **Type**: string
>
> The weekly contact hours in laboratory for a course. This information is outlined in the UTD Course Policies page.
>
> **Example**: 4

> `.offering_frequency`
>
> **Type**: string
>
> The frequency of offering a course. The meanings of each letter can be found in the UTD Course Policies page.
>
> **Example**: "S", "Y", "T", "R"

> `.attributes`
>
> **Type**: Object
>
> Space for additional data describing the course not listed elsewhere.

# Section
## Overview

A `Section` object is the instantiation of a `Course` object with a professor, meeting times, and a grade distribution.

## Object Representation
```ts
Section = {
    "_id": ObjectId,
    "section_number": string,
    "course_reference": ObjectId,
    "section_corequisites": CollectionRequirement,
    "academic_session": AcademicSession,
    "professors": Array<ObjectId>,
    "teaching_assistants": Array<Assistant>,
    "internal_class_number": string,
    "instruction_mode": string,
    "meetings": Array<Meeting>,
    "syllabus_uri": string,
    "grade_distribution": Array<number>,
    "attributes": Object,
}
```

## Properties
> `._id`
> 
> **Type**: ObjectId
> 
> The MongoDB database id for the `Section` object.
>
> **Example**: ObjectId("61ebbb126e3659537e8a14d6")

> `.section_number`
> 
> **Type**: string
> 
> The section's official number.
> 
> **Example**: 002

> `.course_reference`
> 
> **Type**: ObjectId
> 
> An id that points to the course in MongoDB that this section is an instantiation of.
> 
> **Example**:

> `.section_corequisites`
> 
> **Type**: CollectionRequirement
> 
> A collection of all sections that must be taken alongside this section such as specific labs and exam sections.

> `.academic_session`
> 
> **Type**: AcademicSession
> 
> The academic session that the section takes place in.

> `.professors`
> 
> **Type**: Array<ObjectId>
> 
> An array of ids linking to each professor in MongoDB for this section.

> `.teaching_assistants`
> 
> **Type**: Array<Assistant>
> 
> An array of all teaching assistants for this section.

> `.internal_class_number`
> 
> **Type**: string
> 
> The internal (university) number used to reference this section.
> 
> **Example**: 82785

> `.instruction_mode`
> 
> **Type**: string
> 
> The instruction modality for this section.
> 
> **Example**: Traditional

> `.meetings`
> 
> **Type**: Array<Meeting>
> 
> An array of the locations and times that this section meets.

> `.syllabus_uri`
> 
> **Type**: string
> 
> A link to the syllabus on the web.
> 
> **Example**: https://dox.utdallas.edu/syl118093

> `.grade_distribution`
> 
> **Type**: Array<number>
> 
> An array of how many students achieved a certain grade in this section decreating in the order of A+, A, A-, etc.
> 
> **Example**: [2, 3, 5, 4, 3, ...]

> `.attributes`
> 
> **Type**: Object
> 
> Space for additional data describing the section not listed elsewhere.
  
# Exam
## Overview
  
An `Exam` represents a credit-bearing or placement examination at the University of Texas at Dallas.
  
## Object Representation
```ts
ExamType = "AP" | "ALEKS" | "CLEP" | "IB" | "CS placement"
  
abstract Exam = {
  "_id": ObjectId,
  "type": ExamType,
  "yields": Dictionary<int, Array<ObjectId>>,
}
  
APExam extends Exam = {
  "_id": ObjectId,
  "type": "AP",
  "name": string,
  "yields": Dictionary<int, Array<ObjectId>>,
}
  
ALEKSExam extends Exam = {
  "_id": ObjectId,
  "type": "ALEKS",
  "yields": {},
}
  
CLEPExam extends Exam = {
  "_id": ObjectId,
  "type": "CLEP",
  "name": string,
  "yields": Dictionary<int, Array<ObjectId>>,
}
 
IBExam extends Exam = {
  "_id": ObjectId,
  "type": "IB",
  "name": string,
  "level": string,
  "yields": Dictionary<int, Array<ObjectId>>,
}
  
CSPlacementExam extends Exam = {
  "_id": ObjectId,
  "type": "CS placement",
  "yields": {70: [CS1336._id, CS1136._id]},
}
```

## Properties
> `._id`
> 
> **Type**: ObjectId
> 
> The MongoDB database id for the `Exam` object.
>
> **Example**: ObjectId("61ebbb126e3659537e8a14d6")
  
> `.type`
> 
> **Type**: ExamType
> 
> The type of exam object this object represents.
>
> **Example**: AP
  
> `.yields`
> 
> **Type**: Dictionary<int, Array<ObjectId>>
> 
> Maps scores to credit received as a reference to the `Course`. Does not include placement, only actual credit.
>
> **Example**: {}, {70: [CS1336._id, CS1136._id]}
  
> `.name`
> 
> **Type**: string
> 
> The name of the exact exam being taken.
>
> **Example**: "Macroeconomics", "American History I: Early Colonization to 1877", "Environmental Systems
and Societies"
  
> `.level`
> 
> **Type**: string
> 
> The level of the IB exam.
>
> **Example**: Standard, Higher
  
# Academic Session
## Overview

An `AcademicSession` represents the time period in which courses takes place.

## Object Representation
```ts
AcademicSession = {
    "name": string,
    "start_date": string,
    "end_date": string,
}
```
## Properties
> `.name`
>
> **Type**: string
>
> The name of the academic session in question.
>
> **Example**: Spring 2022

> `.start_date`
>
> **Type**: string
>
> The date of classes starting in the academic session.
>
> **Example**: January 18, 2022

> `.end_date`
>
> **Type**: string
>
> The date of classes ending in the academic session.
>
> **Example**: May 5, 2022

# Assistant
## Overview

An 'Assistant' represents a teaching assistant at UT Dallas.

## Object Representation
```ts
Assistant = {
    "first_name": string,
    "last_name": string,
    "role": string,
    "email": string,
}
```

## Properties
> **Type**: ObjectId
>
> The MongoDB database id for the `Assistant` object.
>
> **Example**: ObjectId("61ebbb126e3659537e8a14d6")

> `.first_name`
>
> **Type**: string
>
> The first name of the assistant.
>
> **Example**: John

> `.last_name`
>
> **Type**: string
>
> The last name of the assistant.
>
> **Example**: Doe

> `.role`
>
> **Type**: string
>
> The role of the assistant.
>
> **Example**: "Teaching Assistant", "Guest/Visiting Lecturer"

> `.email`
>
> **Type**: string
>
> The email address to contact the assistant.
>
> **Example**: xxx555555@utdallas.edu
 
# Meeting
## Overview

A 'Meeting' represents a recurring meeting. This schema can represent both recurring meetings and single meetings. Meetings occur repeatedly on the specified days of the week during a period. Non-recurring meetings should have the `start_date` equal to the `end_date`.

## Object Representation
```ts
ModalityType = "pending" | "traditional" | "hybrid" | "flexible" | "remote" | "online"

Meeting = {
    "start_date": string,
    "end_date": string,
    "meeting_days": Array<string>,
    "start_time": string,
    "end_time": string,
    "modality": ModalityType,
    "location": Location,
}
```

## Properties
> `.start_date`
>
> **Type**: string
>
> The start date of a meeting.
>
> **Example**: January 18, 2022

> `.end_date`
>
> **Type**: string
>
> The end date of a meeting.
>
> **Example**: May 5, 2022

> `.meeting_days`
>
> **Type**: Array<string>
>
> A list of all days the meeting occurs during the time period.
>
> **Example**: ["Monday", "Wednesday"]

> `.start_time`
>
> **Type**: string
>
> The time the meeting starts on each meeting day.
>
> **Example**: "10:00am"

> `.end_time`
>
> **Type**: string
>
> The time a meeting ends on each meeting day.
>
> **Example**: "11:15am"

> `.modality`
>
> **Type**: ModalityType
>
> The modality of the meeting following the modality types in [UTD's CourseBook](https://coursebook.utdallas.edu/modalities).
>
> **Example**: traditional

> `.location`
>
> **Type**: Location
>
> The location of the meeting.

# Professor
## Overview

A `Professor` represents a professor employed at the University of Texas at Dallas.

## Object Representation
```ts
Professor = {
    "_id": ObjectId,
    "first_name": string,
    "last_name": string,
    "title": string,
    "email": string,
    "phone_number": string,
    "office": Location,
    "profile_uri": string,
    "office_hours": Array<Meeting>,
}
```

## Properties
> `._id`
>
> **Type**: ObjectId
>
> The MongoDB database id for the `Professor` object.
>
> **Example**: ObjectId("61ebbb126e3659537e8a14d6")

> `.first_name`
>
> **Type**: string
>
> The professor's first name.
>
> **Example**: John

> `.last_name`
>
> **Type**: string
>
> The professor's last name.
>
> **Example**: Doe

> `.title`
>
> **Type**: string
>
> The professor's title.
>
> **Example**: Senior Mathematics Lecturer

> `.email`
>
> **Type**: string
>
> The professor's email address.

> `.phone_number`
>
> **Type**: string
>
> The professor's phone number.

> `.office`
>
> **Type**: Location
>
> The location of the professor's office.

> `.profile_uri`
>
> **Type**: string
>
> A hyperlink pointing to the professor's official university profile.

> `.office_hours`
>
> **Type**: Array<Meeting>
>
> A list of all office hours of the professor.

# Location
## Overview

A location on the UT Dallas campus.

## Object Representation
```ts
Location = {
    "building": string,
    "room": string,
    "map_uri": string,
}
```

## Properties
> `.building`
>
> **Type**: string
>
> The building of the location.
>
> **Example**: "SLC", "ONLINE"

> `.room`
>
> **Type**: string
>
> The room of the location.
>
> **Example**: "2.203", "ONLINE"

> `.map_uri`
>
> **Type**: string
>
> A hyperlink to the UTD room locator.
>
> **Example**: https://locator.utdallas.edu/SLC_2.203

# Requirement
## Overview

A 'Requirement' represents a requirement that can be satisfied. This is probably the most important concept in the API and also has the most cognitive overhead. `Requirement` is abstract and has multiple classes that derive from it to represent the various types requirements at UTD. An overview of these will be given here:

### Course Requirement

A `CourseRequirement` represents a specific `Course` that must be taken to satisfy the requirement. This will represent the majority of requirements.

### Section Requirement

A `SectionRequirement` represents a specific `Section` that must be taken to satisfy the requirement.

### Exam Requirement

An `ExamRequirement` represents a specific `Exam` credit that must be received to satisfy the requirement.

### Major Requirement

A `MajorRequirement` represents a major that a student must be enrolled in to satisfy the requirement.

### Minor Requirement

A `MinorRequirement` represents a minor that a student must be enrolled in to satisfy the requirement.

### GPA Requirement

A `GPARequirement` represents the minimum GPA a student must have to satisfy the requirement.

### Consent Requirement

A `ConsentRequirement` represents the need for consent from a faculty member to satisfy the requirement.

### Collection Requirement

A `CollectionRequirement` represents the need to have some number of requirements from a list of `Requirement`s to satisfy the requirement. This is used to represent groups of requirements, "and" relationships, "some" relationships, and "or" relationships. Collections can contain other collections. Since this is probably the most complicated data structure, I will give some examples here.

For example, a `CollectionRequirement` can contain (inside `options`) the `CourseRequirement` CS 2305 and the `CourseRequirement` CS 2336 with the `required` property set to 2. This represents a requirement that requires both CS 2305 and CS 2336 (i.e. an "and" relationship) to be satisfied. If the `required` property was set to 1, only 1 of CS 2305 and CS 2336 would be required (i.e. an "or" relationship).

As a more real example, consider the prerequisites for the course CS 3305. It requires CE 2305 or CS 2305 or TE 2305 with a grade of C or better and MATH 2414 or MATH 2419. This would be represented as a `CollectionRequirement` with `required` set to 2. The two `options` are two separate `CollectionRequirement`s: one `CollectionRequirement` has a `required` property of 1 with the `options` of the `CourseRequirement`s CE 2305, CS 2305, TE 2305 each with a `minimum_grade` of "C". The other `CollectionRequirement` has a `required` of 1 and the two `options` being the two `CourseRequirement`s MATH 2414 and MATH 2419.

### Hours Requirement

An `HoursRequirement` represents the need to have taken some number of credit hours from a list of `CourseRequirement`s to satisfy the requirement.

### Other Requirement

An `OtherRequirement` represents some miscellaneous need to satisfy the requirement.

## Object Representation
```ts
RequirementType = "course" | "section" | "exam" | "major" | "minor" | "gpa" | "consent" | "collection" | "hours" | "other"

abstract Requirement = {
    "type": RequirementType,
}

CourseRequirement extends Requirement = {
    "type": "course",
    "class_reference": ObjectId,
    "minimum_grade": string,
}

SectionRequirement extends Requirement = {
    "type": "section",
    "section_reference": ObjectId,
}

ExamRequirement extends Requirement = {
    "type": "exam",
    "exam_reference": ObjectId,
    "minimum_score": number,
}

MajorRequirement extends Requirement = {
    "type": "major",
    "major": string,
}

MinorRequirement extends Requirement = {
    "type": "minor",
    "minor": string,
}

GPARequirement extends Requirement = {
    "type": "gpa",
    "minimum": number,
    "subset": string,
}

ConsentRequirement extends Requirement = {
    "type": "consent",
    "granter": string, 
}

OtherRequirement extends Requirement = {
    "type": "other",
    "description": string,
    "condition": string,
}

CollectionRequirement extends Requirement = {
    "type": "collection",
    "name": string,
    "required": number,
    "options": Array<Requirement>,
}

HoursRequirement extends Requirement = {
    "type": "hours",
    "required": number,
    "options": Array<CourseRequirement>,
}
```

## Properties
> `.type`
>
> **Type**: RequirementType
>
> The type of requirement this object represents.
>
> **Example**: collection

> `.class_reference`
>
> **Type**: ObjectId
>
> A reference to the class required in MongoDB in the form of an id.

> `.minimum_grade`
>
> **Type**: string
>
> The minimum grade required for the course to fulfill the requirement. An empty string indicates default behavior.
>
> **Example**: "C", ""

> `.section_reference`
>
> **Type**: ObjectId
>
> A reference to the section required in MongoDB in the form of an id.
>
> **Example**:

> `.exam_reference`
>
> **Type**: ObjectId
>
> A reference to the exam required in MongoDB in the form of an id.
>
> **Example**:

> `.minimum_score`
>
> **Type**: number
>
> The minimum score required on the exam to fulfill the requirement.
>
> **Example**: 70

> `.minimum`
>
> **Type**: number
>
> The minimum GPA required to fulfill the requirement.
>
> **Example**: 3.5

> `.subset`
>
> **Type**: string
>
> The subset of courses that are in question when determining the GPA. If empty, this represents overall GPA.
>
> **Example**: "Major", ""

> `.granter`
>
> **Type**: string
>
> Who grants consent.
>
> **Example**: "Advisor"

> `.description`
>
> **Type**: string
>
> A description of what kind of custom criteria will be used to determine if the requirement is satisfied.

> `.condition`
>
> **Type**: string
>
> The condition given the criteria to determine if the requirement is satisfied.

> `.name`
>
> **Type**: string
>
> A name for the collection to indicate what it holds. May be empty if not very applicable.
>
> **Example**: "Core Curriculum", "", "Major Requirements"

> `.required`
>
> **Type**: number
>
> The minimum number of required requirements necessary to satisfy the requirement.
>
> **Example**: 3

> `.options`
>
> **Type**: Array<Requirement>
>
> A list of all the options that can contribute to satisfying the requirement.
>
> **Example**: [CourseRequirement1, CourseRequirement2, CourseRequirement3, CourseRequirement4, CollectionRequirement1]

# Degree
## Overview

A `Degree` represents either a major, minor, or concentration received from The University of Texas at Dallas.

## Object Representation
```ts
DegreeSubtype = "major" | "minor" | "concentration"

Degree = {
    "_id": ObjectId,
    "subtype": DegreeSubtype,
    "name": string,
    "abbreviation": string,
    "minimum_credit_hours": number,
    "requirements": CollectionRequirement,
}
```

## Properties
> `._id`
>
> **Type**: ObjectId
>
> The MongoDB database id for the `Degree` object.
>
> **Example**: ObjectId("61ebbb126e3659537e8a14d6")

> `.subtype`
>
> **Type**: DegreeSubtype
>
> The subtype of degree that this object represents.
>
> **Example**: Major

> `.name`
>
> **Type**: string
>
> The full name of the degree.
>
> **Example**: Bachelor of Science in Computer Science

> `.abbreviation`
>
> **Type**: string
>
>The abbreviation of the degree.
>
> **Example**: B.S. in Computer Science

> `.minimum_credit_hours`
>
> **Type**: number
>
> The minimum credit hours required for the degree, which can be found on the UTD catalog.
>
> **Example**: 124

> `.requirements`
>
> **Type**: CollectionRequirement
>
> A collection of requirements required to satisfy the degree.

**Reply by jpahm** [reply]
The documentation looks quite clean, though almost all of this typescript is invalid. I would ask that the schemas be at least structurally based on those provided in my post above, even if the names or exact implementation of properties is different.

For some examples of invalid typescript in the above, using `extends` is invalid unless the object in question is a class, which these schemas are not (they should be `type`s). More importantly, statements like
```ts
{
"type": "collection"
}
```
are not valid for our intentions. This does not define a property of name `type` with value `collection`, but instead a property of name `type` with a type of `"collection"`, which is not what we want.

These issues and others were addressed in the raw schemas provided in my post, and I think we should use those along with this documentation.

**Reply by jmyrick02** [reply]
I wrote the schema to be language agnostic since I'm more focusing on the representation of the data within the database. This is almost exactly how it will be represented in MongoDB. The representation in TypeScript will be different. The documentation should inform clearly how our data is organized more than anything else in my opinion. The actual TypeScript code is not necessary for documentation from my understanding.

**Reply by jpahm** [reply]
I understand that, but the schemas also have a purpose in that they're defining the schemas that our backend code uses for actually pushing/pulling this data to/from MongoDB. I do not think presenting the real TypeScript representations of these schemas will detract from the language-agnosticism of the documentation, as it is still up to the front-end language to actually parse these objects.

**Reply by jpahm** [reply]
To further expand on this, my primary concern is that us providing a completely raw "object" representation of these schemas here without any of the finer details does not serve back-end development. I am completely in favor of making the documentation as easy to understand as possible for the front-end, and I see your intention, but I think that this documentation should be a true representation of the schemas so as to not compromise its usefulness to back-end.

**Reply by jmyrick02** [reply]
This is the database schema. The database (MongoDB) does not utilize TypeScript at all. What you are suggesting should be a separate documentation. We could create a discussion to start this more internal-facing documentation. I have found some basic examples that show a database schema that is similar to what is here.

https://upload.wikimedia.org/wikipedia/commons/9/94/MediaWiki_1.28.0_database_schema.svg
https://drawsql.app/templates/koel

**Reply by jpahm** [reply]
I am aware that MongoDB does not utilize TypeScript, I believe there is a miscommunication here.

To rephrase my concern, I do not like how this is presenting the schemas as some sort of psuedo-typescript with `extends` and other TypeScript language constructs. We should either lean fully into presenting the schemas as they will be in the database (i.e., raw objects with no inter-type relationships) or present them as our internal schemas types, which would be in the form of TypeScript.

**Reply by jmyrick02** [reply]
I think the problem with representing the objects as raw objects with no inter-type relationships is that it gets difficult to describe some things. For example, the `CollectionRequirement` type has the property `options: Array<Requirement>`. This means that any object in the array has to be of type Requirement. The `extends` keyword on `GPARequirement`, `ClassRequirement`, etc. clearly represents and conveys this relationship and that it can be included in the `options` property of the `CollectionRequirement`. Not denoting this could make it unclear or ambiguous to the end-user.

I'm open to using syntax other than `extends` to denote inheritance relations but it seems pretty clear and comprehensible and is common to other languages such as Java so isn't TypeScript exclusive.

Edit: I do think that it's not clear that the base classes not meant to be initialized in the database so perhaps we should add an `abstract` keyword to `Requirement` and `Exam`.

**Reply by charliemahana** [reply]
When you guys finalize the schemas, please use something similar to drawsql which @jmyrick02 linked above. However, do not save this documentation as an image. We need to be able to change it later as we get more data and the scope of what the API provides may change.

**Reply by TrystonMinsquero** [reply]
I think the current schema is readable by people who are not familiar with typescript and gets the point across which is what we want. I sympathize with Josh as this will be useful, so @hochladen post the schema using proper typescript.

**Reply by jpahm** [reply]
I will concede that it is helpful to be able to see the type relationships, which is one of the reasons I was trying to advocate for presenting the TypeScript representations of the objects, as one of the main purposes of TypeScript is to allow for effectively typing JSON (which is precisely what MongoDB will be storing). TypeScript's syntax makes the types, relationships, and structure of objects clear as far as both implementation and object representation are concerned.

That being said, I would be willing to compromise so long as we are in agreement that the TypeScript representations need clear documentation *somewhere*, which we seem to be.

**Reply by jpahm** [reply]
Overall, I'm proud of what we've been able to develop over the past few days, and think this will be a major step in the right direction. I will write up a full typescript-formatted version of this documentation tomorrow, as I do not think it's immediately necessary (and I'm tired lol). 

Good work everyone :)

**Comment by AdamMcAdamson** [comment]
Athena has expressed a desire for API data aggregation and to enable data over time analysis.

Data over time analysis can be resolved with, and aggregating around, the 'academic_session' properity of sections. **We do not handle 8 week sessions well at the moment!!**

They also want aggregates for Courses to Sections. Professors to Courses and Professors to Sections.

I propose the following additions to the schema:
```typescript
CourseSection = {
    "course": Course,
    "academic_session_name": string,
    "sections": Array<Section>
}

ProfessorCourses = {
   "professor": Professor,
   "courses": Array<CourseSection>
}
```

We may want to swap for ObjectId's, idk..  have to get to class so this is somewhat half-baked, I will finish this later :)

**Reply by jpahm** [reply]
Perhaps I misunderstand, but I think this could be more cleanly done by simply expanding the current `professor` schema to contain references to all of that professor's sections (something we had planned earlier, but never made it into the current schema). Courses already have references to their sections, and sections have references to their course, so any of the other aggregates should already be able to be handled using the existing schema infrastructure. 

As far as handling 8-week sessions go, I do not see what you mean in that we do not handle them well. The `AcademicSession` schema only specifies start/end dates, so it should be fully capable of representing academic sessions of any length equally.

**Reply by jmyrick02** [reply]
I like @hochladen's idea. It's more elegant than the solution Athena proposed. However, I think we should check with Athena to check if this implementation would be ok.

I further propose that the professor object would hold an array of references to all sections taught during their academic career and I think this array should be sorted in descending order with respect to the `end_date` of the section's `AcademicSession` object so Athena can find sections in a specific time period.

This solution would require more work on Athena's end but would be better on our end and probably more extendable to other projects using our API.

**Reply by jpahm** [reply]
In addition to this, if we want to cut down how much work the front-end has to do, we could always handle a good chunk of the actual database parsing (i.e. grabbing sections from a specific date range) on the API end and just pass the front-end what they need.

**Reply by jmyrick02** [reply]
I completely overlooked that. In that case, I completely agree with this idea and I will propose a change to the schema and some endpoints.

**Reply by AdamMcAdamson** [reply]
Unless I'm completely missing something, Courses don't currently have references to their sections, right? I think we all agree that they should.

I was under the assumption that the start and end dates had some flexibility within a given session, though I think I am mistaken. There is no distinction between 8-week sessions other than the dates, which will require some processing overhead with us handling date ranges to resolve. I figure since the data is there it may make sense to include it.

Professor's having reference to sections (an courses?) they teach and building endpoints where the server handles aggregation and lookups would work. That is what I initially assumed they were asking for, and am probably confused in thinking otherwise.

**Reply by jmyrick02** [reply]
I may have misled you @AdamMcAdamson if I misunderstood Eric and Charlie's suggestion. I apologize if that's the case. Courses having references to their sections could be useful, although currently I believe no other Nebula projects would have a direct use case for that. It may be justified in adding since it would probably come up as a requested feature eventually.

**Reply by jpahm** [reply]
@AdamMcAdamson You're correct about courses not currently referencing their sections. I had proposed such in some of my schema proposals, and still think there's pretty much no downside to including them (i.e. by adding a `section_references` property to the `Course` schema) given that we have the ability to do so.

**Reply by AdamMcAdamson** [reply]
Okay, then yeah, I think we should do that too.

I will include in the endpoints discussion a [potential solution](https://github.com/UTDNebula/nebula-api/discussions/59#discussioncomment-2329308) for handling aggregation and lookups.

**Reply by jpahm** [reply]
Also, upon doing more research about the usage of MongoDB's `_id` field, it appears that we are not actually constrained to using the `ObjectID` type for the IDs. `_id` can optionally take on any type, or we can generate `ObjectID` types using hexadecimal strings (i.e. the course/section IDs that Coursebook provides us). This should let us have a very powerful referencing system, as MongoDB should be able to reference courses and sections by the same IDs that Coursebook gives us, rather than us needing to do any sort of manual hashing or other workarounds.

**Reply by AdamMcAdamson** [reply]
As noted in #82: 
There are a few edge cases not properly represented with the Exam schema/data, as with AP Physics where for "Physics Majors: Credit for 2326 + 2126 is applied only if PHYS 3416 is passed with a grade of C or higher; otherwise you earn 4 SCH of free electives".
Most are AP credit specific, and are outlined by the endnotes on the [AP Credit website](https://oue.utdallas.edu/undergraduate-advising/ap-credit).

Also, there are a few 'or' conditions across the exams, which are not handled (AP Math and 2 IB Math exams).

Any comments on how/if we should handle them would be greatly appreciated.

**Comment by ZzRanger** [comment]
There are some requirements in certain Major degree plans that are more difficult to validate. Below are two instances of this:

**History major requirement**
<img width="605" alt="Screen Shot 2022-03-09 at 10 46 26 PM" src="https://user-images.githubusercontent.com/47403443/157591495-4b6a7b95-fc99-46ac-8386-92f270d3b5c7.png">
 
**Economics and Finance Double Major Requirement**
<img width="665" alt="Screen Shot 2022-03-09 at 10 48 31 PM" src="https://user-images.githubusercontent.com/47403443/157591699-d5c933da-ee0b-4f7c-b110-91b1385fdda8.png">

To handle these kinds of edge cases, I propose adding the following properties to CollectionRequirements:
1. "selfValidate": boolean
2. "description": string (optional field)

The "selfValidate" field denotes whether or not the user will have to validate that their degree plan satisfies the following CollectionRequirement. This will be set to "true" whenever edge cases requirements are encountered.

The "description" field will contain information about how to fulfill a certain CollectionRequirement. This will only be used for edge cases. Below is an example description from the Economics and Finance example.

> "A practicum experience is required; the student has the option of zero to 3 semester credit hours, depending on the particular internship, and preference for additional electives."

Please let me know what you guys think about the proposed change.

**Reply by jmyrick02** [reply]
We currently have the `OtherRequirement` which should satisfy anything requiring a description or handling edge cases. In this case a boolean flag would not be required since encountering the `OtherRequirement` would inherently imply that self-validation is required.

**Reply by ZzRanger** [reply]
Okay cool that makes sense; that should work on the Planner side. Thanks!
